pinecone/index_connection.go

```diff
@@ -0,0 +1,358 @@
+package pinecone
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"github.com/pinecone-io/go-pinecone/internal/gen/data"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"log"
+)
+
+type IndexConnection struct {
+	apiKey     string
+	dataClient *data.VectorServiceClient
+	grpcConn   *grpc.ClientConn
+}
+
+func newIndexConnection(apiKey string, host string) (*IndexConnection, error) {
+	config := &tls.Config{}
+	target := fmt.Sprintf("%s:443", host)
+	conn, err := grpc.Dial(
+		target,
+		grpc.WithTransportCredentials(credentials.NewTLS(config)),
+		grpc.WithAuthority(target),
+		grpc.WithBlock(),
+	)
+
+	if err != nil {
+		log.Fatalf("fail to dial: %v", err)
+		return nil, err
+	}
+
+	dataClient := data.NewVectorServiceClient(conn)
+
+	idx := IndexConnection{apiKey: apiKey, dataClient: &dataClient, grpcConn: conn}
+	return &idx, nil
+}
+
+func (idx *IndexConnection) Close() error {
+	err := idx.grpcConn.Close()
+	return err
+}
+
+type UpsertVectorsRequest struct {
+	Vectors   []*Vector
+	Namespace string
+}
+
+func (idx *IndexConnection) UpsertVectors(ctx *context.Context, in *UpsertVectorsRequest) (uint32, error) {
+	vectors := make([]*data.Vector, len(in.Vectors))
+	for i, v := range in.Vectors {
+		vectors[i] = vecToGrpc(v)
+	}
+
+	req := &data.UpsertRequest{
+		Vectors:   vectors,
+		Namespace: in.Namespace,
+	}
+
+	res, err := (*idx.dataClient).Upsert(idx.akCtx(*ctx), req)
+	if err != nil {
+		return 0, err
+	}
```

Comment:
**thought:** I'm not sure what the errors that are returned from the gRPC calls look like, but we may need to eventually think about centralized mapping / wrapping of things if that makes sense in Go.

Reply:
In Go, every error is an `error`. Unless we wanted different messages, these are generally easier to maintain as pass-through.